docs/proposals/20210427-unitedtopology.md

```diff
@@ -0,0 +1,256 @@
+    - `Adaptive` means choose the right subset by sequence of subsets and also enough resources
+  - `adaptive`:
+    - `disableSimulationSchedule`: whether to disable the simulation schedule in webhook
+    - `rescheduleCriticalSeconds`: the critical duration that controller will trigger a Pod to reschedule to another subset
+
+**status**:
+- `subsetStatuses`: statuses of subsets defined in spec
+  - `missingReplicas`: the missing number of this subset, most of the time it is (maxReplicas - the number of active Pods)
+  - `creating`: the map of Pods that just allocated to this subset by webhook
+  - `deleting`: the map of Pods been deleted or evicted that just noticed by webhook
+  - `missingReschedule`: the number of Pods that should be rescheduled to this subset
+
+### Requirements
+
+UnitedTopology requires Pod webhook to inject rules into Pods and notice the deletion and eviction of Pods.
+
+So if the `PodWebhook` feature-gate is set to `false`, UnitedTopology will also be disabled.
+
+### Implementation Details/Notes/Constraints
+
+UnitedTopology have both webhook and controller.
+
+#### Replicas control
+
+When a Pod is creating and its workload has a related UnitedTopology,
+webhook will check the subsetStatuses in the UnitedTopology one by one:
+
+1. for each subset, calculate the `delta` number by `missingReplicas - len(creating) + len(deleting)`
+2. if `delta` > 0, choose this subset temporarily
+   1. try to update subsetStatus and put this pod into `creating` map
+   2. if it conflicts, get the UnitedTopology again and go back to step 1
+   3. if it succeeds, inject the rules of this subset into the Pod
+3. if `delta` <= 0, go to the next subset
+4. if there is no available subset left, just let the Pod go
+
+Also, when a Pod that belongs to a subset is deleting or evicting, webhook will put it into the `deleting` map in subsetStatus.
+
+#### Deletion priority
+
+Since Kubernetes 1.21, there is a new annotation definition on Pod:
+
+```yaml
+    controller.kubernetes.io/pod-deletion-cost: -1
+```
+
+(In 1.21, users need to enable `PodDeletionCost` feature-gate, and since 1.22 it will be enabled by default)
+
+The integer value is the deletion cost of the Pod. A Pod without this annotation will be considered as deletion-cost `0`.
+
+When ReplicaSet(Deployment) tries to scale in, it will sort its Pods by following sequence:
+
+1. Pods that have not been scheduled
+2. Pods with `Pending` and `Unknown` phases
+3. Pods in not-ready state
+4. Pods with **lower deletion-cost**
+5. Pods that have deployed on the same Node
+6. Pods with higher restart count
+7. Pods that have newer creationTimestamp
+8. ...
+
+CloneSet will also support this annotation to implement deletion priority.
+
+And then, UnitedTopology will define three levels for the Pods belongs to subsets:
+
+1. `Guaranteed` for Pods that in the `maxReplicas` of its subset, deletion-cost will be set bigger than `0`
```

Comment:
consider the case that deployments with idc and cloud instance, we may prefer to delete cloud instance even without maxReplicas